src/main/java/com/conveyal/taui/models/ModificationStop.java

```diff
@@ -144,29 +175,32 @@ private ModificationStop(Coordinate c, String id, double distanceFromStart) {
         return stopDwellTimes;
     }
 
-    static int[] getHopTimes (List<ModificationStop> stops, int[] segmentSpeeds) {
+    static int[] getHopTimes (List<ModificationStop> stops, int[] segmentSpeedsKph) {
         if (stops == null || stops.size() < 2) {
             return new int[0];
         }
 
-        int[] hopTimes = new int[stops.size() - 1];
-
+        int[] hopTimesSeconds = new int[stops.size() - 1];
         ModificationStop lastStop = stops.get(0);
+
+        // This "real" stop index is the index into the originally supplied stops, ignoring the auto-generated ones.
         int realStopIndex = 0;
-        for (int i = 0; i < hopTimes.length; i++) {
+        for (int i = 0; i < hopTimesSeconds.length; i++) {
             ModificationStop stop = stops.get(i + 1);
             double hopDistance = stop.distanceFromStart - lastStop.distanceFromStart;
 
-            int segmentSpeed = segmentSpeeds.length > realStopIndex ? segmentSpeeds[realStopIndex] : DEFAULT_SEGMENT_SPEED;
-            hopTimes[i] = (int) (hopDistance / (segmentSpeed * 1000) * 3000);
+            int segmentSpeedKph = segmentSpeedsKph.length > realStopIndex ? segmentSpeedsKph[realStopIndex] :
+                    DEFAULT_SEGMENT_SPEED_KPH;
+            hopTimesSeconds[i] = (int) (hopDistance / (segmentSpeedKph * METERS_PER_KM) * SECONDS_PER_HOUR);
 
-            if (stop.id != null) {
+            if (stop.autoGenerated) {
```

Comment:
For future reference, this line is the main fix for #174.  The realStopIndex, used when assigning segment-specific speeds to hop times, was not being incremented at newly created (i.e. not snapped to existing) stops.  This meant that in cases with speeds that varied by segment, the speeds were potentially applied to the wrong segments.

Reply:
Oh.  Shouldn't this be `if (!stop.autoGenerated)`?  The realStopIndex should be incremented if the stop is a realStop (i.e. not auto generated).